src/Microsoft.Framework.PackageManager/List/LoadableAssemblyWalker.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.Versioning;
+using Microsoft.Framework.PackageManager.Algorithm;
+using Microsoft.Framework.Runtime;
+using NuGet;
+
+namespace Microsoft.Framework.PackageManager.List
+{
+    public class LoadableAssemblyWalker
+    {
+        private readonly static FrameworkName AspNetCore50 = VersionUtility.ParseFrameworkName("aspnetcore50");
+
+        private readonly GacDependencyResolver _gacResolver;
+        private readonly FrameworkName _framework;
+        private readonly ApplicationHostContext _hostContext;
+        private readonly IDictionary<Library, LibraryDescription> _libDictionary;
+        private readonly string _runtimeFolder;
+
+        public LoadableAssemblyWalker(
+            FrameworkName framework,
+            ApplicationHostContext hostContext,
+            IDictionary<Library, LibraryDescription> libDictionary,
+            string runtimeFolder)
+        {
+            _framework = framework;
+            _hostContext = hostContext;
+            _runtimeFolder = runtimeFolder;
+            _libDictionary = libDictionary;
+
+            _gacResolver = new GacDependencyResolver();
+        }
+
+        public ISet<string> Walk(IGraphNode<Library> root)
+        {
+            var assemblies = new HashSet<string>();
+            var libraries = new HashSet<Library>();
+            root.DepthFirstPreOrderWalk(visitNode: (node, _) => VisitLibrary(node, _, libraries, assemblies));
+
+            return assemblies;
+        }
+
+        private bool VisitLibrary(IGraphNode<Library> node, IEnumerable<IGraphNode<Library>> ancestors, ISet<Library> visitedLibraries, ISet<string> assemblies)
+        {
+            if (visitedLibraries.Add(node.Item))
+            {
+                LibraryDescription desc;
+                if (_libDictionary.TryGetValue(node.Item, out desc))
+                {
+                    foreach (var loadableAssembly in desc.LoadableAssemblies)
+                    {
+                        DepthFirstGraphTraversalExtensions.DepthFirstPreOrderWalk(
```

Comment:
Though, actually, now that I look at it a bit closer, I'm not sure I'm a fan of it being an extension method. It's effectively an extension method on `object`, which is a kind of code smell. I think it's unlikely for this extension method to make sense on every single type in the entire world. It has a rather specific case in which it works. Perhaps either a regular static method or some other design would be a better fit.

https://github.com/aspnet/Home/wiki/Engineering-guidelines#extension-method-patterns